File: apple/QuenderinKit/Sources/QuenderinKit/AgentTool.swift
import Foundation

/// A capability the agent can invoke during its loop. On-device and offline —
/// math, text utilities, local search, etc.
///
/// > Deliberately NOT "drive another app": iOS sandboxes one app from
/// > controlling another, so the agent acts through *tools* it owns, not by
/// > pixel-driving the OS. That's the honest, shippable shape of "autonomous"
/// > on iOS.
public protocol AgentTool: Sendable {
    /// Stable identifier the planner emits, e.g. "calculator".
    var name: String { get }
    /// One line the planner sees to decide when to use it.
    var purpose: String { get }
    /// Run with a string input, returning an observation string.
    func run(_ input: String) async throws -> String
}

/// Returns its input verbatim — a deterministic tool for tests and as the
/// simplest possible example.
public struct EchoTool: AgentTool {
    public init() {}
    public let name = "echo"
    public let purpose = "Repeat the input back. Use to restate something."
    public func run(_ input: String) async throws -> String { input }
}

/// Evaluates a simple arithmetic expression (`+ - * /`, parentheses, unary
/// minus, decimals). Uses a hand-rolled recursive-descent parser — NOT
/// NSExpression, which raises uncatchable ObjC exceptions on malformed input.
public struct CalculatorTool: AgentTool {
    public init() {}
    public let name = "calculator"
    public let purpose = "Evaluate arithmetic like \"12 * (3 + 4)\", \"2^10\", or \"sqrt(16)\" (+ - * / ^ %, parentheses, sqrt/abs/floor/ceil, pi/e)."

    public func run(_ input: String) async throws -> String {
        guard let value = ArithmeticParser.evaluate(input) else {
            return "Couldn't evaluate \"\(input)\"."
        }
        // Render integers without a trailing .0
        if value.rounded() == value && abs(value) < 1e15 {
            return String(Int64(value))
        }
        return String(value)
    }
}

/// A tiny, safe arithmetic evaluator. Returns nil on any malformed input rather
/// than crashing — important because the input comes from an LLM.
enum ArithmeticParser {
    static func evaluate(_ text: String) -> Double? {
        var parser = Parser(tokens: tokenize(text))
        guard let value = parser.parseExpression(), parser.isAtEnd, value.isFinite else { return nil }
        return value   // reject NaN/Inf (e.g. (-2)^0.5, 2^9999) → "Couldn't evaluate", not "nan"/"inf"
    }

    private static func tokenize(_ text: String) -> [String] {
        var tokens: [String] = []
        var number = ""
        var ident = ""   // a function name (sqrt) or constant (pi) — runs of letters
        func flush() {
            if !number.isEmpty { tokens.append(number); number = "" }
            if !ident.isEmpty { tokens.append(ident); ident = "" }
        }
        for char in text where !char.isWhitespace {
            if char.isNumber || char == "." {
                if !ident.isEmpty { flush() }        // letter→digit boundary
                number.append(char)
            } else if char.isLetter {
                if !number.isEmpty { flush() }        // digit→letter boundary
                ident += String(char).lowercased()
            } else {
                flush()
                if "+-*/^%()".contains(char) { tokens.append(String(char)) } else { return [] }
            }
        }
        flush()
        return tokens
    }

    private struct Parser {
        let tokens: [String]
        var pos = 0
        var isAtEnd: Bool { pos >= tokens.count }
        private func peek() -> String? { pos < tokens.count ? tokens[pos] : nil }
        private mutating func advance() -> String? { defer { pos += 1 }; return peek() }

        // A recursion-depth cap: a Swift stack overflow is NOT a catchable Error, so deeply
        // nested parens (e.g. adversarial model output) would hard-crash the process instead of
        // failing the calc gracefully. Bail to nil past this depth (C4; mirrors the Kotlin twin).
        private static let maxDepth = 100

        // expression = term (('+' | '-') term)*
        mutating func parseExpression(_ depth: Int = 0) -> Double? {
            if depth > Self.maxDepth { return nil }
            guard var value = parseTerm(depth + 1) else { return nil }
            while let op = peek(), op == "+" || op == "-" {
                _ = advance()
                guard let rhs = parseTerm(depth + 1) else { return nil }
                value = (op == "+") ? value + rhs : value - rhs
            }
            return value
        }

        // term = factor (('*' | '/' | '%') factor)*
        private mutating func parseTerm(_ depth: Int) -> Double? {
            if depth > Self.maxDepth { return nil }
            guard var value = parseFactor(depth + 1) else { return nil }
            while let op = peek(), op == "*" || op == "/" || op == "%" {
                _ = advance()
                guard let rhs = parseFactor(depth + 1) else { return nil }
                switch op {
                case "/": guard rhs != 0 else { return nil }; value /= rhs
                case "%": guard rhs != 0 else { return nil }; value = value.truncatingRemainder(dividingBy: rhs)
                default:  value *= rhs
                }
            }
            return value
        }

        // factor = '-' factor | power   (unary minus binds LOOSER than '^', so -2^2 = -(2^2) = -4)
        private mutating func parseFactor(_ depth: Int) -> Double? {
            if depth > Self.maxDepth { return nil }
            if peek() == "-" { _ = advance(); guard let f = parseFactor(depth + 1) else { return nil }; return -f }
            return parsePower(depth + 1)
        }

        // power = primary ('^' factor)?   (right-associative, binds TIGHTER than '*'/'/': 2^3^2 = 2^9,
        // 2*3^2 = 2*9; the RHS recurses through factor so a unary/negative exponent like 2^-1 works)
        private mutating func parsePower(_ depth: Int) -> Double? {
            if depth > Self.maxDepth { return nil }
            guard let base = parsePrimary(depth + 1) else { return nil }
            guard peek() == "^" else { return base }
            _ = advance()
            guard let exponent = parseFactor(depth + 1) else { return nil }
            return pow(base, exponent)
        }

        // Parity-safe single-arg functions + constants — IDENTICAL across Swift/Kotlin/JS. Deliberately
        // NOT round / log / ln / sin / cos / tan: their half-rounding + domain/precision behavior differs
        // across stdlibs, which would re-introduce the cross-platform divergences this project keeps fixing.
        private static let functions: [String: @Sendable (Double) -> Double] = [
            "sqrt": { Foundation.sqrt($0) }, "abs": { Swift.abs($0) },
            "floor": { Foundation.floor($0) }, "ceil": { Foundation.ceil($0) },
        ]
        private static let constants: [String: Double] = ["pi": .pi, "e": M_E]

        // primary = number | constant | function '(' expression ')' | '(' expression ')'
        private mutating func parsePrimary(_ depth: Int) -> Double? {
            if depth > Self.maxDepth { return nil }
            guard let token = peek() else { return nil }
            if token == "(" {
                _ = advance()
                guard let value = parseExpression(depth + 1), peek() == ")" else { return nil }
                _ = advance()
                return value
            }
            if let fn = Self.functions[token] {                       // function call: name '(' expr ')'
                _ = advance()
                guard peek() == "(" else { return nil }
                _ = advance()
                guard let arg = parseExpression(depth + 1), peek() == ")" else { return nil }
                _ = advance()
                return fn(arg)
            }
            if let constant = Self.constants[token] { _ = advance(); return constant }
            if let number = Double(token) { _ = advance(); return number }
            return nil
        }
    }
}

File: android/quenderin-core/src/main/kotlin/ai/quenderin/core/AgentTool.kt
package ai.quenderin.core

import kotlin.math.abs
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.math.pow
import kotlin.math.sqrt

/**
 * A capability the agent can invoke during its loop. On-device and offline — math, text
 * utilities, local search, etc. Synchronous here (the Kotlin core is coroutine-free); the
 * app wraps long-running tools in a coroutine. Mirrors iOS `AgentTool`.
 *
 * Deliberately NOT "drive another app": the OS sandboxes apps from controlling each other,
 * so the agent acts through *tools* it owns — the honest, shippable shape of "autonomous".
 */
interface AgentTool {
    /** Stable identifier the planner emits, e.g. "calculator". */
    val name: String
    /** One line the planner sees to decide when to use it. */
    val purpose: String
    /** Run with a string input, returning an observation string. */
    fun run(input: String): String
}

/** Returns its input verbatim — deterministic, the simplest possible example. */
class EchoTool : Capability {
    override val name = "echo"
    override val purpose = "Repeat the input back. Use to restate something."
    override fun run(input: String): String = input
}

/** Evaluates simple arithmetic. Uses a hand-rolled recursive-descent parser that returns
 *  null on malformed input rather than throwing — the input comes from an LLM. */
class CalculatorTool : Capability {
    override val name = "calculator"
    override val purpose = "Evaluate arithmetic like \"12 * (3 + 4)\", \"2^10\", or \"sqrt(16)\" (+ - * / ^ %, parentheses, sqrt/abs/floor/ceil, pi/e)."
    override fun run(input: String): String {
        val value = ArithmeticParser.evaluate(input) ?: return "Couldn't evaluate \"$input\"."
        return if (value == Math.rint(value) && abs(value) < 1e15) value.toLong().toString() else NumberRender.canonical(value)
    }
}

/**
 * ONE number→string rendering for tool observations, byte-identical across the twins — the
 * platform-native shortest-repr strings diverge in digits AND presentation ("1.18…E21" here,
 * "1.18…e+21" on iOS; plain vs sci thresholds differ too), feeding the agent loop different
 * observations for the same value (twin-drift audit, agent-session P2). Contract: 12 significant
 * digits with HALF-EVEN ties (BigDecimal here; C's correctly-rounded %.11e on the Swift twin —
 * probed to agree, tie case included), plain decimal for exponents -5…14, canonical scientific
 * ("1.18059162072e+21") outside.
 */
object NumberRender {
    fun canonical(v: Double): String {
        if (!v.isFinite()) return v.toString()
        val bd = java.math.BigDecimal(v).round(java.math.MathContext(12, java.math.RoundingMode.HALF_EVEN))
        val negative = bd.signum() < 0
        val unscaled = bd.unscaledValue().abs().toString()
        val exponent = unscaled.length - 1 - bd.scale()
        var digits = unscaled.trimEnd('0')
        if (digits.isEmpty()) digits = "0"
        return assemble(if (negative) "-" else "", digits, exponent)
    }

    /** The shared presentation half — same logic, token for token, as the Swift twin. */
    internal fun assemble(sign: String, digits: String, exponent: Int): String {
        if (exponent in -5..14) {
            if (exponent >= digits.length - 1) {
                return sign + digits + "0".repeat(exponent - (digits.length - 1))
            }
            if (exponent >= 0) {
                return sign + digits.substring(0, exponent + 1) + "." + digits.substring(exponent + 1)
            }
            return sign + "0." + "0".repeat(-exponent - 1) + digits
        }
        val mantissa = if (digits.length == 1) digits else "${digits.first()}.${digits.substring(1)}"
        return "$sign${mantissa}e${if (exponent >= 0) "+" else "-"}${kotlin.math.abs(exponent)}"
    }
}

/** Tiny, safe arithmetic evaluator (`+ - * /`, parentheses, unary minus, decimals). */
object ArithmeticParser {
    fun evaluate(text: String): Double? {
        val tokens = tokenize(text) ?: return null
        val parser = Parser(tokens)
        val value = parser.parseExpression() ?: return null
        // reject NaN/Inf (e.g. (-2)^0.5, 2^9999) → "Couldn't evaluate", not "NaN"/"Infinity"
        return if (parser.isAtEnd && value.isFinite()) value else null
    }

    private fun tokenize(text: String): List<String>? {
        val tokens = mutableListOf<String>()
        var number = StringBuilder()
        var ident = StringBuilder()   // a function name (sqrt) or constant (pi) — runs of letters
        fun flush() {
            if (number.isNotEmpty()) { tokens.add(number.toString()); number = StringBuilder() }
            if (ident.isNotEmpty()) { tokens.add(ident.toString()); ident = StringBuilder() }
        }
        for (c in text) {
            if (c.isWhitespace()) continue
            when {
                c.isDigit() || c == '.' -> { if (ident.isNotEmpty()) flush(); number.append(c) }
                c.isLetter() -> { if (number.isNotEmpty()) flush(); ident.append(c.lowercaseChar()) }
                else -> {
                    flush()
                    if (c in "+-*/^%()") tokens.add(c.toString()) else return null
                }
            }
        }
        flush()
        return tokens
    }

    private class Parser(val tokens: List<String>) {
        var pos = 0
        val isAtEnd: Boolean get() = pos >= tokens.size
        private fun peek(): String? = tokens.getOrNull(pos)
        private fun advance() { pos++ }

        // expression = term (('+' | '-') term)*
        fun parseExpression(depth: Int = 0): Double? {
            if (depth > MAX_DEPTH) return null  // cap recursion — parity with the Swift twin (C4)
            var value = parseTerm(depth + 1) ?: return null
            while (true) {
                val op = peek()
                if (op != "+" && op != "-") break
                advance()
                val rhs = parseTerm(depth + 1) ?: return null
                value = if (op == "+") value + rhs else value - rhs
            }
            return value
        }

        // term = factor (('*' | '/' | '%') factor)*
        private fun parseTerm(depth: Int): Double? {
            if (depth > MAX_DEPTH) return null
            var value = parseFactor(depth + 1) ?: return null
            while (true) {
                val op = peek()
                if (op != "*" && op != "/" && op != "%") break
                advance()
                val rhs = parseFactor(depth + 1) ?: return null
                when (op) {
                    "/" -> { if (rhs == 0.0) return null; value /= rhs }
                    "%" -> { if (rhs == 0.0) return null; value %= rhs }
                    else -> value *= rhs
                }
            }
            return value
        }

        // factor = '-' factor | power   (unary minus binds LOOSER than '^', so -2^2 = -(2^2) = -4)
        private fun parseFactor(depth: Int): Double? {
            if (depth > MAX_DEPTH) return null
            if (peek() == "-") { advance(); val f = parseFactor(depth + 1) ?: return null; return -f }
            return parsePower(depth + 1)
        }

        // power = primary ('^' factor)?   (right-associative, binds TIGHTER than '*'/'/': 2^3^2 = 2^9,
        // 2*3^2 = 2*9; the RHS recurses through factor so a unary/negative exponent like 2^-1 works)
        private fun parsePower(depth: Int): Double? {
            if (depth > MAX_DEPTH) return null
            val base = parsePrimary(depth + 1) ?: return null
            if (peek() != "^") return base
            advance()
            val exponent = parseFactor(depth + 1) ?: return null
            return base.pow(exponent)
        }

        // primary = number | constant | function '(' expression ')' | '(' expression ')'
        private fun parsePrimary(depth: Int): Double? {
            if (depth > MAX_DEPTH) return null
            val token = peek() ?: return null
            if (token == "(") {
                advance()
                val value = parseExpression(depth + 1) ?: return null
                if (peek() != ")") return null
                advance()
                return value
            }
            val fn = FUNCTIONS[token]
            if (fn != null) {                            // function call: name '(' expr ')'
                advance()
                if (peek() != "(") return null
                advance()
                val arg = parseExpression(depth + 1) ?: return null
                if (peek() != ")") return null
                advance()
                return fn(arg)
            }
            CONSTANTS[token]?.let { advance(); return it }
            val number = token.toDoubleOrNull() ?: return null
            advance()
            return number
        }

        private companion object {
            const val MAX_DEPTH = 100
            // Parity-safe single-arg functions + constants — IDENTICAL across Swift/Kotlin/JS. NOT
            // round/log/ln/sin/cos/tan (their half-rounding + domain/precision behavior differs across
            // stdlibs, which would re-introduce the cross-platform divergences this project keeps fixing).
            val FUNCTIONS: Map<String, (Double) -> Double> = mapOf(
                "sqrt" to { x: Double -> sqrt(x) }, "abs" to { x: Double -> abs(x) },
                "floor" to { x: Double -> floor(x) }, "ceil" to { x: Double -> ceil(x) },
            )
            val CONSTANTS: Map<String, Double> = mapOf("pi" to Math.PI, "e" to Math.E)
        }
    }
}
